R/Methods-FAKinGroupResults.R

```diff
@@ -217,85 +217,100 @@ setMethod("runSimulation", "FAKinGroupResults", function(object, nsim=50000,
+                if(length(expectedAffInKin) >= 2){
+                    ## that's the slowest part of the function... subsetting of the
+                    ## kinship matrix...
+                    ## we get some speedup if we subset a matrix, not a Matrix!
+                    expectedAffInKin <- match(expectedAffInKin,
+                                              colnames(kinPhenotyped))
+                    expectedKin <- kinPhenotyped[expectedAffInKin, expectedAffInKin]
+                    diag(expectedKin) <- NA
+                    ## get the table with the frequencies of kinship values...
+                    expectedKinFreqs <- table(as.vector(expectedKin))
+                    observedKinFreqs <- z$kinfreq
+                    if(length(expectedKinFreqs) > 0 & length(observedKinFreqs) > 0){
+                        ## names of the table are the kinship values, ordered increasingly by size.
+                        expKinShipValue <- names(expectedKinFreqs[length(expectedKinFreqs)])
+                        obsKinShipValue <- names(observedKinFreqs[length(observedKinFreqs)])
+                        if(as.numeric(expKinShipValue) >= as.numeric(obsKinShipValue)){
+                            ## if the kinship value is the same, compare the counts.
+                            if(expKinShipValue==obsKinShipValue){
+                                if(expectedKinFreqs[expKinShipValue] >=
+                                   observedKinFreqs[obsKinShipValue])
+                                    Res[2] <- 1
+                            }else{
+                                ## means it's larger, so add +1
                                 Res[2] <- 1
-                        }else{
-                            ## means its larger, so add +1
-                            Res[2] <- 1
+                            }
                         }
                     }
                 }
-            }
-            return(Res)
-        }))
-        return(ExpRes)
-    })
-    ## BigRes is now a list, length nsim, each element being a matrix, ncol=2,
-    ## nrow=length affectedKins.
-    ExpRatio <- colSums(do.call(rbind, lapply(BigRes,
-                                              function(simRes){simRes[, 1]})))
-    ExpKin <- colSums(do.call(rbind, lapply(BigRes,
-                                            function(simRes){simRes[, 2]})))
-    ExpVals <- do.call(rbind, lapply(BigRes, function(simRes){simRes[, 3]}))
-    ExpVals <- split(t(ExpVals), f=names(affectedKins))
-    expDensity <- lapply(ExpVals, density)
-    expHist <- lapply(ExpVals, hist, plot=FALSE)
-    expDensity <- expDensity[names(affectedKins)]
-    expHist <- expHist[names(affectedKins)]
-    names(ExpRatio) <- names(affectedKins)
-    names(ExpKin) <- names(affectedKins)
+                return(Res)
+            }))
+            ExpRatio <- ExpRatio + ExpRes[, 1]
+            ExpKin   <- ExpKin   + ExpRes[, 2]
+            BigRes <- append(BigRes, list(ExpRes[, 3]))
```

Comment:
If we created the `BigRes` list already with the right size above we could here use simply: `BigRes[[i]] <- ExpRes[, 3]` - this *should* avoid costly copying of the whole object because a single element is added. I'm however not sure if `list(ExpRes[, 3])` or `ExpRes[, 3]` should be added - I think `ExpRes[, 3]` should suffice, but that depends also how you access the data from `BigRes` below.